### PR TITLE
addJVHelpers utility functions added.

### LINKED ISSUE #23: followRelationships to include related object at top level instead of _jv.rels
Currently followRelationships includes related objects under _jv.rels which is not very "beatiful". Especially, if there will be multiple levels of relations:
article._jv.rels.comments._jv.rels.author

According to JSON:API specs, attributes and relations are both fields and there must not be attribute and relation with the same name (https://jsonapi.org/format/#document-resource-object-fields).

I'd like to suggest injecting related objects already at top level of the object, which should be safe, thus making access to nested objects more easy to read and understand:
article.comments.author

### PR DESCRIPTION
**Pull request by mrichar1**
Fixes #23 

Code adds utility functions to an existing object's `_jv` child object to allow `rels` and `attrs` introspection, as suggested in #23.

Tests are failing (as the root now contains relationships! :smile: ) so the test data will need modified to expect these.

@staam let me know if this looks like what you were thinking of, and if so I'll update the tests(and add new ones for`addJvHelpers` (which probably needs a better name!)

### CODE REVIEW

File: src/jsonapi-vuex.js [hunk 1]
```diff
@@ -359,6 +359,37 @@ const jsonapiModule = (api, conf = {}) => {
 
 // Helper methods
 
+const addJvHelpers = (obj) => {
+  // Add Utility functions to _jv child object
+  Object.assign(obj[jvtag], {
+    is_rel(name) {
```
Comment (mhluska) [inline src/jsonapi-vuex.js]: Should we use camel case for these?

File: src/jsonapi-vuex.js [hunk 2]
```diff
@@ -359,6 +359,37 @@ const jsonapiModule = (api, conf = {}) => {
...
+      return name in obj[jvtag]['relationships']
```
Comment (mhluska) [inline src/jsonapi-vuex.js]: Is this testing presence in an object? I think the recommended way is `[hasOwnProperty](https://stackoverflow.com/a/455366)`.

File: src/jsonapi-vuex.js [hunk 3]
```diff
@@ -359,6 +359,37 @@ const jsonapiModule = (api, conf = {}) => {
...
+    },
+    is_attr(name) {
+      return name != '_jv' && !this.is_rel(name)
```
Comment (mhluska) [inline src/jsonapi-vuex.js]: Minor nit: if you don't need to do type coercion, it's better to use [!==](https://stackoverflow.com/a/359547)

File: src/jsonapi-vuex.js (hunk 1 above)
Comment (mrichar1) [reply src/jsonapi-vuex.js]: haha - old habits die hard! Quite right!

File: src/jsonapi-vuex.js (hunk 3 above)
Comment (mrichar1) [reply src/jsonapi-vuex.js]: Agreed - will fix this.

**Comment by staaam** [comment]
Yeah, looks pretty same I've meant!

File: src/jsonapi-vuex.js [hunk 4]
```diff
@@ -359,6 +359,37 @@ const jsonapiModule = (api, conf = {}) => {
...
 // Helper methods
 
+const addJvHelpers = (obj) => {
+  // Add Utility functions to _jv child object
+  Object.assign(obj[jvtag], {
+    isRel(name) {
+      return obj[jvtag]['relationships'].hasOwnProperty(name)
+    },
+    isAttr(name) {
+      return name !== '_jv' && !this.isRel(name)
```
Comment (staaam) [inline src/jsonapi-vuex.js]: replace '_jv' with jvtag ?

Comment (staaam) [inline src/jsonapi-vuex.js]:
```suggestion
      return name !== jvtag && !this.isRel(name)
```

File: src/jsonapi-vuex.js [hunk 5]
```diff
@@ -359,6 +359,37 @@ const jsonapiModule = (api, conf = {}) => {
...
+        if (obj[jvtag]['relationships'].hasOwnProperty(key)) {
+          rel[key] = val
+        }
+      }
+      return rel
+    },
+    get attrs() {
+      const att = {}
+      for (let [key, val] of Object.entries(obj)) {
+        if (key != jvtag && !(obj[jvtag]['relationships'].hasOwnProperty(key))) {
```
Comment (staaam) [inline src/jsonapi-vuex.js]:
```suggestion
        if (key !== jvtag && !(obj[jvtag]['relationships'].hasOwnProperty(key))) {
```

File: src/jsonapi-vuex.js [hunk 6]
```diff
@@ -359,6 +359,37 @@ const jsonapiModule = (api, conf = {}) => {
 
 // Helper methods
 
+const addJvHelpers = (obj) => {
+  // Add Utility functions to _jv child object
+  Object.assign(obj[jvtag], {
+    isRel(name) {
+      return obj[jvtag]['relationships'].hasOwnProperty(name)
```
Comment (staaam) [inline src/jsonapi-vuex.js]: worth taking out obj[jvtag]['relationships'].hasOwnProperty(name) to some local isRel or something, or reusing isRel in other calls - it's used 3 times in this snippet

File: src/jsonapi-vuex.js (hunk 4 above)
Comment (mrichar1) [reply src/jsonapi-vuex.js]: Fixed!

File: src/jsonapi-vuex.js (hunk 6 above)
Comment (mrichar1) [reply src/jsonapi-vuex.js]: Have set other the calls to reuse `isRel`

**Comment by mrichar1** [comment]
So there are a couple of extra changes (and questions) relating to this PR:

* Now that `rels` can be in the root, they need to be removed when creating JSONAPI (basically #36 again but in a different form).

* Do we want to keep the `rels` in the object when it is stored? I think this might be a good idea, because then actions and getters return the 'same' formatted object. We also currently call `checkAndFollowRelationships` in the `get` getter anyway, so this would be more efficient. Thoughts?

**Comment by mrichar1** [comment]
I'm hitting some interesting issues with trying to update the test suite for this PR.

There are 2 issues I can see - 1 obvious, 1 less so:

1. The first is that the test data doesn't contain the helper methods as added by `addJvHelper` so some of the deep comparisons fail. This is easily fixed by using `addJvHelper` in the test data construction (having thoroughly tested it directly elsewhere!). However...

2. The circular nature of the test data is breaking things. Things are fine in master where `_jv.rels` is an object, but they break here where `_jv.rels` is a getter - presumably something to do with how `chai`'s `deep.equal` spots recursion? (In either case, doing `JSON.stringify(_jv.rels)` complains about a cyclic dependency error in both cases, so something clever is happening somewhere in the test suite).

The simple option I'm going to start with is to just do a shallow compare (i.e. that the relationship keys end up in the root) and avoid recursing into the rel objects - though there may be some benefit to doing a small amount of digging if just the keys isn't enough.

Ideally we should also add more formal recursion tests to go along with the changes introduced by #28 since this is the kind of area where regression and bugs are most likely to end up - but that's probably a separate issue.

**Comment by mrichar1** [comment]
OK - that's things mostly fleshed out and working.

A few things still to do:

1. Add tests for `addJvHelpers`
2. (Possibly) remove relationships in `normToStore` (see above).
3. (Possibly) move all the `addJvHelpers` calls from specific tests into the fixtures (make this default behaviour).

**Comment by mrichar1** [comment]
Some interesting issues popping up with this one...

1. Using `getters` in the object results in them appearing to 'cache' the original data in the object, so `patch` and `delete don't work. Changing from `getters` to simple functions fixes this. Don't know enough about JS to know *why* this is happening, but it's not an issue to switch.

2. `post` needs `normToJsonapi` to not try to use `attrs` as a user-provided object won't have the `jvAddHelpers` utility functions attached. `isRel` needs protection against empty objects.

3. `delete` isn't working any more. Something about the relationships is keeping references and objects don't actually disappear when deleted. This can be seen in the testapp - delete `1` and `3` and nothing happens until you delete `2` at which point all three objects disappear from the store.

Have pushed the code which fixes 1 and 2 (and demonstrates 3)

**Comment by mhluska** [comment]
@mrichar1 for your first point, the use of `Object.assign` is what is causing caching. Example:

```js
class Cat {
}

const cat = new Cat();

Object.assign(cat, {
  get info() {
    return {
      type: 'minx',
    };
  }
});

console.log(cat.info === cat.info); // Prints true
```

```js
class Cat {
  get info() {
    return {
      type: 'minx',
    };
  }
}

const cat = new Cat();

console.log(cat.info === cat.info); // Prints false
```

**Comment by staaam** [comment]
wow... it "grew" a little with all the required changes to make this work.. I hope it worth it :)

**Comment by mrichar1** [comment]
@mhluska Ahh! that makes perfect sense, once you think about it! :+1: 

I'll move back to getters (using `defineProperty`) and tidy up the code so that rels don't end up in the store, then see where we are with the deletion issue.

**Comment by mrichar1** [comment]
OK - that's the `Object.assign`/`getter` issue resolved, and I've modified `normToStore` to remove rels from root (and tidied it up a bit).

this hasn't fixed the delete issue in the `e2e` tests, but I've dug a bit deeper and the behaviour is a bit strange.

If you delete any item in the UI, then the `delete` action is successfully deleting the items from the vuex store (seen in the `devtools` panel).

However, if you delete items `2` or `3` (in either order) they also disappear from the UI (`widgets` computed property) but if you delete `1` it only disappears when `2` and `3` are also gone. This implies the issue is with the computed property not being 're-computed' in some way? I'll dig deeper, but happy for any insights!

**Comment by mrichar1** [comment]
Turns out there was a obvious bug in the `get` getter - if you specified a non-existent `id` (i.e. `1` after it has been deleted) then instead of returning `{}` it returned the whole collection for that `type`. Fixed (and added tests!) and the tests are all passing. hopefully there aren't any more of those kinds of issue hiding in the code!

File: src/jsonapi-vuex.js [hunk 7]
```diff
@@ -359,6 +364,41 @@ const jsonapiModule = (api, conf = {}) => {
...
+  // Add Utility functions to _jv child object
+  Object.assign(obj[jvtag], {
+    isRel(name) {
+      return get(obj, [jvtag, 'relationships'], {}).hasOwnProperty(name)
+    },
+    isAttr(name) {
+      return name !== jvtag && !this.isRel(name)
+    },
+  })
+  Object.defineProperty(obj[jvtag], 'rels', {
```
Comment (staaam) [inline src/jsonapi-vuex.js]: worth adding comment why rels/attr added with defineProperty (maybe with link to current discussion)

**Comment by staaam** [comment]
That looks really great!

File: src/jsonapi-vuex.js (hunk 7 above)
Comment (mrichar1) [reply src/jsonapi-vuex.js]: Good point - it's the kind of thing that looks like it needs tidy information, which will introduce a regression. Will update with a comment.

**Comment by mrichar1** [comment]
Right - that's (I think) the last of the tests added (directly testing `addJvHelpers`) and a minor fix added (`isAttr` should return False for non-existent attributes).

I think it's now ready to merge but will leave it for a while in case there's any last minute feedback you'd like to add?

After that I'll merge, fix the new minor Issues, then release a new version. I think it should probably be a major version bump, since this is a significant change!

**Comment by mrichar1** [comment]
OK - that's this code merged now - I'll run through the other Issues/PRs and tidy up any simple ones, then release a new major version.

After that we can revisit the other enhancements (and wait for the inevitable bugs that this PR will have snuck in!)

**Pull request merged**